### ISSUE TITLE
Is there a v2 branch yet? 

### ISSUE THREAD

**Issue by trusktr** [JSMonk/hegel#323]
I'd like to start learning, and if there's a v2 branch, I'd like to check it out.

This issue may be better off in [Discussions](https://docs.github.com/en/discussions/quickstart).

**Comment by JSMonk** [comment]
No, currently it is in separate Repo because it will be rewritten from scratch.

**Comment by CrazyPython** [comment]
For projects, how compatible is it with v1?

**Comment by JSMonk** [comment]
It will be a superset of the first version, so it will have backward compatibility.

**Comment by trusktr** [comment]
Idea: make it compilable to WebAssembly! Is there an opportunity to work with @dcode from AssemblyScript perhaps? Maybe Hegel can generate ASTs (in a way that doesn't have the downside of TypeScript transforms), and AssemblyScript can use them to map to its Wasm output?

**Comment by trusktr** [comment]
Oops, wrong guy, I meant @dcodeIO.

**Comment by MaxGraey** [comment]
@CrazyPython It very depends on task. See another examples:
https://github.com/acutmore/type-off
https://twitter.com/matthewcp/status/1390013392950284291

**Comment by MaxGraey** [comment]
> but you can write the same fast code in JavaScript, if you know what to avoid.

That's not really true. Ofc you could do something [like this](https://medium.com/leaningtech/a-javascript-optimizing-compiler-3fd3f49bd071) in theory. Or you could use d8/chrome with `--trace-deopt` to figure outing de-optimizations paths and polymorphic code but this [really hard](https://mrale.ph/blog/2018/02/03/maybe-you-dont-need-rust-to-speed-up-your-js.html). Also as I mentioned some runtimes (builtins) methods / classes like string concatenation you can't speedup at all.

**Comment by CrazyPython** [comment]
Thanks for the links.

WASM and JS use the same optimizing backend. You can manipulate both to get similar results.

> Also as I mentioned some runtimes (builtins) methods / classes like string concatenation you can't speedup at all.

My guess is this JS code is just as fast. Benchmark this on V8 and JSC.
```
function toList(arr: string[]): string {
  for (let i = 0, len = arr.length; i < len; i++) {
    yield* arr;
  }
}
```

Either way, if you want a fast compiler, it comes from writing a character-by-character streaming compiler and avoiding IRs and complex data structures whenever possible. You can do that in any language.

**Comment by MaxGraey** [comment]
> WASM and JS use the same optimizing backend. You can manipulate both to get similar results.

Yes and no. First of all JS hasn't some operations like popcount. [See how this affect to performance](https://medium.com/@mbebenita/webassembly-is-30x-faster-than-javascript-c71ea54d2f96). The reason is very [simple](https://godbolt.org/z/sohsKrvzv). Also turbofun have special optimizations only for wasm backend which not yet enabled. Like this one: https://bugs.chromium.org/p/v8/issues/detail?id=11298&q=unroll&can=2. And still not apply some important optimisations for i64/u64: https://bugs.chromium.org/p/v8/issues/detail?id=11085&q=wasm%20optimize%20i64&can=2

**Comment by texastoland** [comment]
Seems like that discussion got really off-topic. @JSMonk How long do you anticipate until a v2 beta?

**Comment by JSMonk** [comment]
Hi @texastoland,
I planed to implement the MVP in September/October of this year.

**Comment by hazelTheParrot** [comment]
"I planed to implement the MVP in September/October of this year."

How is the progress? Soon october now. So a v. 2.0 will be released when?

**Comment by texastoland** [comment]
Looks like @JSMonk [started working on Kotlin](https://twitter.com/rage_monk/status/1429835691748364294?s=20) (congrats!) so probably delayed ⏳

**Comment by JSMonk** [comment]
Hi @hazelTheParrot. Sure, I started working on it, and hope (not sure about the October), but we will present MVP soon, with a pretty cool news.

**Comment by boogerlad** [comment]
don't mean to rush you and know estimates are hard. Do you think it'll be sometime this year?

**Comment by JSMonk** [comment]
I'm working hard to finish the MVP this year, but I'm not sure.
